lib/api/apiUtils/object/abortMultipartUpload.js

```diff
@@ -105,9 +106,39 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
                     }
                     cb();
                 });
-            }, () => next(null, mpuBucket, storedParts, destBucket));
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
         },
-        function deleteMetadata(mpuBucket, storedParts, destBucket, next) {
+        function deleteObjectData(mpuBucket, storedParts, destBucket, objectMD, next) {
+            // Filtering parts that has already been delete by the previous step
+            let partLocations = storedParts.map(item => item.value.partLocations);
+            partLocations = [].concat(...partLocations);
+            partLocations = new Set(partLocations.map(loc => loc.key));
+            const objectLocationLeft = objectMD.location.filter(loc => !partLocations.has(loc.key));
+
+            return async.eachLimit(objectLocationLeft, 5, (loc, cb) => {
+                data.delete(loc, log, err => {
+                    if (err) {
+                        log.fatal('delete object data failed', { err });
+                    }
+                    cb();
+                });
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
+        },
+        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD, next) {
```

Comment:
If I'm not mistaken I think it's better to delete the metadata first, to avoid leaving the object in an inconsistent state (i.e. unreadable), especially if there are failures.

Reply:
correct me if I'm wrong, but if we crash for some reason after deleting the metadata first, we will end up with orphans never recoverable nor deletable.. ?

Reply:
From what I understand, usually we prefer having orphans in the storage layer rather than having an orphan in the metadata (orphan in the sense that we cannot do anything with it, and the data is not accessible). Both options are not perfect, but from the client point of view it allows to have a clean state. Nowadays we run tools to cleanup such orphans, but later we may be smarter and queue these operations internally...